cloudflare/resource_cloudflare_waiting_room.go

```diff
@@ -0,0 +1,223 @@
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"session_duration": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Second),
+			Update: schema.DefaultTimeout(30 * time.Second),
+		},
+	}
+}
+
+func buildWaitingRoom(d *schema.ResourceData) cloudflare.WaitingRoom {
+	return cloudflare.WaitingRoom{
+		Name:                  d.Get("name").(string),
+		Description:           d.Get("description").(string),
+		Suspended:             d.Get("suspended").(bool),
+		Host:                  d.Get("host").(string),
+		Path:                  d.Get("path").(string),
+		TotalActiveUsers:      d.Get("total_active_users").(int),
+		NewUsersPerMinute:     d.Get("new_users_per_minute").(int),
+		CustomPageHtml:        d.Get("custom_page_html").(string),
+		SessionDuration:       d.Get("session_duration").(int),
+		QueueAll:              d.Get("queue_all").(bool),
+		DisableSessionRenewal: d.Get("disable_session_renewal").(bool),
+	}
+}
+
+func resourceCloudflareWaitingRoomReadFromWaitingRoomId(d *schema.ResourceData, meta interface{}, zoneID, waitingRoomID string) error {
+	client := meta.(*cloudflare.API)
+	waitingRoom, err := client.WaitingRoom(context.Background(), zoneID, waitingRoomID)
+	if err != nil {
+		if strings.Contains(err.Error(), "HTTP status 404") {
+			log.Printf("[WARN] Removing waiting room from state because it's not found in API")
+			d.SetId("")
+			return nil
+		}
+		name := d.Get("name").(string)
+		return fmt.Errorf("Error getting waiting room %q: %s", name, err)
+	}
+	d.SetId(waitingRoom.ID)
+	d.Set("name", waitingRoom.Name)
+	d.Set("description", waitingRoom.Description)
+	d.Set("suspended", waitingRoom.Suspended)
+	d.Set("host", waitingRoom.Host)
+	d.Set("path", waitingRoom.Path)
+	d.Set("queue_all", waitingRoom.QueueAll)
+	d.Set("new_users_per_minute", waitingRoom.NewUsersPerMinute)
+	d.Set("total_active_users", waitingRoom.TotalActiveUsers)
+	d.Set("session_duration", waitingRoom.SessionDuration)
+	d.Set("disable_session_renewal", waitingRoom.DisableSessionRenewal)
+	d.Set("custom_page_html", waitingRoom.CustomPageHtml)
+	return nil
+}
+
+func resourceCloudflareWaitingRoomCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+
+	newWaitingRoom := buildWaitingRoom(d)
+
+	waitingRoom, err := client.CreateWaitingRoom(context.Background(), zoneID, newWaitingRoom)
+
+	if err != nil {
+		name := d.Get("name").(string)
+		return fmt.Errorf("Error creating waiting room %q: %s", name, err)
+	}
+
+	d.SetId(waitingRoom.ID)
+
+	return resourceCloudflareWaitingRoomRead(d, meta)
+}
+
+func resourceCloudflareWaitingRoomRead(d *schema.ResourceData, meta interface{}) error {
+	waitingRoomID := d.Id()
+	zoneID := d.Get("zone_id").(string)
+
+	return resourceCloudflareWaitingRoomReadFromWaitingRoomId(d, meta, zoneID, waitingRoomID)
```

Comment:
this functionality should be pulled into the `Read` method

Reply:
Done